igor-web/src/main/groovy/com/netflix/spinnaker/igor/jenkins/JenkinsBuildMonitor.groovy

```diff
@@ -134,123 +133,87 @@ class JenkinsBuildMonitor implements PollingMonitor {
-                                                                || (build.number == lastBuild && cachedBuild.lastBuildBuilding != build.building)) {
-                                                            log.info  "sending event for intermediate build ${build.number}"
-                                                            postEvent(echoService, cachedBuilds, oldProject, master)
-                                                        }
-                                                    } catch (e) {
-                                                        log.error("An error occurred fetching ${master}:${project.name}:${build.number}", e)
-                                                    }
-                                                }
-                                            }
-                                        } catch (e) {
-                                            log.error("failed getting builds for ${master}", e)
-                                        }
-                                    }
-                                }
-                            } else {
-                                log.info "New Build: ${master}: ${project.name} : ${project.lastBuild.number} : " +
-                                        "${project.lastBuild.result}"
-                                addToCache = true
-                                cache.setLastBuild(master, project.name, project.lastBuild.number, project.lastBuild.building)
-                            }
-                            if (addToCache) {
-                                project.lastBuild.result = project?.lastBuild?.result ?: project.lastBuild.building ? BUILD_IN_PROGRESS : ""
-                                log.debug "setting result to ${project.lastBuild.result}"
-                                postEvent(echoService, cachedBuilds, project, master)
-                                results << [previous: cachedBuild, current: project]
-                            }
-                        } catch (e) {
-                            log.error("fail processing build : ${project?.name}", e)
+                    // 2. post events for finished builds
+                    completedBuilds.forEach { build ->
+                        Boolean eventPosted = cache.getEventPosted(master, job.name, cursor, build.number)
+                        if (!eventPosted) {
+                            log.info("posting build event for build ${master}:${job.name}:${build.number}")
+                            postEvent(echoService, new Project(name: job.name, lastBuild: build), master)
+                            cache.setEventPosted(master, job.name, cursor, build.number)
                         }
                     }
-            )
-
-            log.info("Took ${System.currentTimeMillis() - startTime}ms to retrieve projects (master: ${master})")
 
+                    // 3. advance cursor when all builds have completed in the interval
```

Comment:
It feels like this logic could be simpler if you set the cursor to be the highest timestamp across all builds instead of the timestamp of the last build. This way you don't have to keep track of builds that are still in progress and can look at only completed builds.